applications/GeoMechanicsApplication/custom_constitutive/coulomb_with_tension_cut_off_impl.cpp

```diff
@@ -104,51 +46,106 @@ bool CoulombWithTensionCutOffImpl::IsAdmissibleSigmaTau(const Vector& rTrialSigm
-
-        if (IsStressAtTensionCutoffReturnZone(rTrialSigmaTau, mTensionCutOff.GetTensileStrength(),
-                                              apex, corner_point)) {
-            return ReturnStressAtTensionCutoffReturnZone(
-                rTrialSigmaTau, mTensionCutOff.DerivativeOfFlowFunction(rTrialSigmaTau),
-                mTensionCutOff.GetTensileStrength());
+        if (IsStressAtTensionCutoffReturnZone(rTrialSigmaTau)) {
+            return ReturnStressAtTensionCutoffReturnZone(rTrialSigmaTau);
         }
 
-        if (IsStressAtCornerReturnZone(
-                rTrialSigmaTau, mCoulombYieldSurface.DerivativeOfFlowFunction(rTrialSigmaTau, AveragingType),
-                corner_point)) {
-            result = corner_point;
+        if (IsStressAtCornerReturnZone(rTrialSigmaTau, AveragingType)) {
+            result = CalculateCornerPoint();
         } else { // Regular failure region
-            result = ReturnStressAtRegularFailureZone(
-                rTrialSigmaTau, mCoulombYieldSurface,
-                mCoulombYieldSurface.DerivativeOfFlowFunction(rTrialSigmaTau, AveragingType));
+            result = ReturnStressAtRegularFailureZone(rTrialSigmaTau, AveragingType);
         }
 
-        const auto lambda = mCoulombYieldSurface.CalculatePlasticMultiplier(
-            rTrialSigmaTau, mCoulombYieldSurface.DerivativeOfFlowFunction(rTrialSigmaTau, AveragingType));
-        const double kappa = kappa_start + mCoulombYieldSurface.CalculateEquivalentPlasticStrain(
-                                               rTrialSigmaTau, AveragingType, lambda);
+        const auto kappa = kappa_start + mCoulombYieldSurface.CalculateEquivalentPlasticStrainIncrement(
+                                             rTrialSigmaTau, AveragingType);
         mCoulombYieldSurface.SetKappa(kappa);
 
-        double error = std::abs(mCoulombYieldSurface.YieldFunctionValue(result));
-        if (error < mCoulombYieldSurface.GetConvergenceTolerance()) break;
+        if (std::abs(mCoulombYieldSurface.YieldFunctionValue(result)) < mAbsoluteYieldFunctionValueTolerance) {
+            break;
+        }
     }
     return result;
 }
 
+Vector CoulombWithTensionCutOffImpl::CalculateCornerPoint() const
+{
+    // Check whether the tension cut-off lies beyond the apex
+    auto result                 = Vector{ZeroVector(2)};
+    result[0]                   = mCoulombYieldSurface.CalculateApex();
+    const auto tensile_strength = mTensionCutOff.GetTensileStrength();
+    if (tensile_strength > result[0]) return result;
+
+    const auto cohesion = mCoulombYieldSurface.GetCohesion();
+    const auto sin_phi  = std::sin(mCoulombYieldSurface.GetFrictionAngleInRadians());
+    const auto cos_phi  = std::cos(mCoulombYieldSurface.GetFrictionAngleInRadians());
+    result[0]           = (tensile_strength - cohesion * cos_phi) / (1.0 - sin_phi);
+    result[1]           = (cohesion * cos_phi - tensile_strength * sin_phi) / (1.0 - sin_phi);
+    return result;
```

Comment:
By using `UblasUtilities::CreateVector` we can make the code slightly more elegant:
```suggestion
    const auto tensile_strength = mTensionCutOff.GetTensileStrength();
    if (const auto apex = mCoulombYieldSurface.CalculateApex(); tensile_strength > apex)
        return UblasUtilities::CreateVector({apex, 0.0});

    const auto cohesion = mCoulombYieldSurface.GetCohesion();
    const auto sin_phi  = std::sin(mCoulombYieldSurface.GetFrictionAngleInRadians());
    const auto cos_phi  = std::cos(mCoulombYieldSurface.GetFrictionAngleInRadians());
    return UblasUtilities::CreateVector({(tensile_strength - cohesion * cos_phi) / (1.0 - sin_phi),
                                         (cohesion * cos_phi - tensile_strength * sin_phi) / (1.0 - sin_phi)});
```
Note that the early exit now looks consistent with the result returned at line 129.